node/src/components/block_synchronizer/signature_acquisition.rs

```diff
@@ -109,3 +109,317 @@ impl SignatureAcquisition {
+        let (public_0, secret_0) = validators.get(0).unwrap();
+        let finality_signature =
+            FinalitySignature::create(block_hash, era_id, secret_0, public_0.clone());
+        assert_matches!(
+            signature_acquisition.apply_signature(finality_signature, &weights),
+            Acceptance::NeededIt
+        );
+        assert_equal(signature_acquisition.have_signatures(), [public_0]);
+        assert_equal(signature_acquisition.not_vacant(), [public_0]);
+        assert!(signature_acquisition.have_no_vacant() == false);
+        assert_equal(
+            signature_acquisition.not_pending(),
+            validators.iter().map(|(p, _)| p),
+        );
+        assert_eq!(
+            signature_acquisition.signature_weight(),
+            SignatureWeight::Insufficient
+        );
+
+        // Signature for the validator #2 weighting 3:
+        let (public_2, secret_2) = validators.get(2).unwrap();
+        let finality_signature =
+            FinalitySignature::create(block_hash, era_id, secret_2, public_2.clone());
+        assert_matches!(
+            signature_acquisition.apply_signature(finality_signature, &weights),
+            Acceptance::NeededIt
+        );
+        assert_equal(
+            signature_acquisition.have_signatures(),
+            [public_0, public_2],
+        );
+        assert_equal(signature_acquisition.not_vacant(), [public_0, public_2]);
+        assert!(signature_acquisition.have_no_vacant() == false);
+        assert_equal(
+            signature_acquisition.not_pending(),
+            validators.iter().map(|(p, _)| p),
+        );
+        // The total signed weight is 4/10, which is higher than 1/3:
+        assert_eq!(
+            signature_acquisition.signature_weight(),
+            SignatureWeight::Weak
+        );
+
+        // Signature for the validator #3 weighting 4:
+        let (public_3, secret_3) = validators.get(3).unwrap();
+        let finality_signature =
+            FinalitySignature::create(block_hash, era_id, secret_3, public_3.clone());
+        assert_matches!(
+            signature_acquisition.apply_signature(finality_signature, &weights),
+            Acceptance::NeededIt
+        );
+        assert_equal(
+            signature_acquisition.have_signatures(),
+            [public_0, public_2, public_3],
+        );
+        assert_equal(
+            signature_acquisition.not_vacant(),
+            [public_0, public_2, public_3],
+        );
+        assert!(signature_acquisition.have_no_vacant() == false);
+        assert_equal(
+            signature_acquisition.not_pending(),
+            validators.iter().map(|(p, _)| p),
+        );
+        // The total signed weight is 8/10, which is higher than 2/3:
+        assert_eq!(
+            signature_acquisition.signature_weight(),
+            SignatureWeight::Strict
+        );
+    }
+
+    #[test]
+    fn apply_signature_has_the_expected_behavior_low_finality() {
```

Comment:
Nit:
I suggest naming the tests more precisely, ie. avoid generic statements like "expected behavior". Maybe extract the common code from both "apply_signature" tests (which would be nearly the entire test) and make the common code accept the Ratio parameter.

Then you can have two tests:
```
should_return_weak_when_have_1_and_1_10_is_required(Ratio::new(1, 10))
should_return_insufficient_when_have_1_and_1_3_is_required(Ratio::new(1, 3))
```
or something along these lines

Reply:
Done